File: smart_generator.py
"""
Smart Pattern-Based Card Generator
Learns from real working cards and generates high-quality validation cards
with diverse CVV/expiry combinations
"""
import json
import random
import os
from pathlib import Path
from typing import List, Dict, Optional
from collections import defaultdict
from datetime import datetime


class PatternData:
    """Stores pattern information for a specific 8-digit prefix"""
    def __init__(self, prefix: str):
        self.prefix = prefix
        self.suffix_digits = [[] for _ in range(8)]  # Observed digits at each position
        self.months = []
        self.years = []
        self.cvvs = []
        self.bank = ""
        self.country = ""
        self.card_type = ""
        self.level = ""
        self.count = 0


class SmartGenerator:
    """Pattern-based card generator that learns from real cards"""
    
    def __init__(self, cards_file: str = "output/cards.json"):
        self.cards_file = Path(cards_file)
        self.patterns: Dict[str, PatternData] = {}
        self.all_cvvs: List[str] = []
        self.all_months: List[str] = []
        self.all_years: List[str] = []
        
    def load_cards(self) -> List[Dict]:
        """Load cards from JSON file"""
        cards = []
        
        if not self.cards_file.exists():
            print(f"[!] Cards file not found: {self.cards_file}")
            return cards
        
        try:
            with open(self.cards_file, 'r', encoding='utf-8') as f:
                data = json.load(f)
                if isinstance(data, list):
                    for card in data:
                        cn = card.get('card_number', '')
                        if len(cn) >= 15:
                            cards.append(card)
        except Exception as e:
            print(f"[X] Error loading cards: {e}")
        
        return cards
    
    def build_patterns(self, cards: List[Dict]):
        """Build pattern database from real cards"""
        cvv_set = set()
        month_set = set()
        year_set = set()
        
        for card in cards:
            cn = card.get('card_number', '')
            if len(cn) < 16:
                continue
            
            prefix = cn[:8]
            suffix = cn[8:]
            
            # Initialize pattern if new
            if prefix not in self.patterns:
                self.patterns[prefix] = PatternData(prefix)
            
            p = self.patterns[prefix]
            p.count += 1
            
            # Record suffix digits at each position
            for i, ch in enumerate(suffix[:8]):
                if ch.isdigit():
                    p.suffix_digits[i].append(int(ch))
            
            # Record month, year, cvv
            month = card.get('expiry_month')
            year = card.get('expiry_year')
            cvv = card.get('cvv', '')
            
            if month:
                month_str = f"{month:02d}" if isinstance(month, int) else str(month).zfill(2)
                p.months.append(month_str)
                month_set.add(month_str)
            
            if year:
                year_str = str(year)
                p.years.append(year_str)
                year_set.add(year_str)
            
            if cvv and len(cvv) >= 3:
                p.cvvs.append(cvv)
                cvv_set.add(cvv)
            
            # Set metadata
            if not p.bank:
                p.card_type = card.get('card_type', '')
        
        # Store global distributions
        self.all_cvvs = list(cvv_set)
        self.all_months = list(month_set)
        self.all_years = list(year_set)
        
        # Filter to patterns with 2+ cards
        self.patterns = {k: v for k, v in self.patterns.items() if v.count >= 2}
        
        print(f"[OK] Built {len(self.patterns)} patterns from {len(cards)} cards")
        print(f"[OK] Global: {len(self.all_cvvs)} CVVs, {len(self.all_months)} months, {len(self.all_years)} years")
    
    def calculate_luhn_check(self, partial: str) -> int:
        """Calculate Luhn check digit"""
        total = 0
        is_second = True
        
        for ch in reversed(partial):
            d = int(ch)
            if is_second:
                d *= 2
                if d > 9:
                    d -= 9
            total += d
            is_second = not is_second
        
        return (10 - (total % 10)) % 10
    
    def make_luhn_valid(self, card_number: str) -> str:
        """Make card number Luhn-valid"""
        partial = card_number[:-1]
        check_digit = self.calculate_luhn_check(partial)
        return partial + str(check_digit)
    
    def generate_from_pattern(self, pattern: PatternData) -> Dict:
        """Generate a card from a specific pattern"""
        # Determine card length
        first_digit = pattern.prefix[0] if pattern.prefix else '4'
        if first_digit == '3':  # AMEX
            card_len = 15
            cvv_len = 4
        else:
            card_len = 16
            cvv_len = 3
        
        suffix_len = card_len - len(pattern.prefix)
        
        # Generate suffix based on observed digit patterns
        suffix = ""
        for pos in range(suffix_len - 1):  # -1 for Luhn digit
            if pos < len(pattern.suffix_digits) and pattern.suffix_digits[pos]:
                # Pick random digit from observed digits at this position
                digit = random.choice(pattern.suffix_digits[pos])
                suffix += str(digit)
            else:
                suffix += str(random.randint(0, 9))
        suffix += "0"  # Placeholder for Luhn
        
        # Build card number
        card_num = pattern.prefix + suffix
        card_num = self.make_luhn_valid(card_num)
        
        # === DIVERSE CVV GENERATION ===
        r = random.random()
        if r < 0.30 and pattern.cvvs:
            cvv = random.choice(pattern.cvvs)
        elif r < 0.70 and self.all_cvvs:
            cvv = random.choice(self.all_cvvs)
        else:
            cvv = ''.join([str(random.randint(0, 9)) for _ in range(cvv_len)])
        
        # Ensure correct length
        cvv = cvv.zfill(cvv_len)[:cvv_len]
        
        # === DIVERSE MONTH GENERATION ===
        r = random.random()
        if r < 0.25 and pattern.months:
            month = random.choice(pattern.months)
        elif r < 0.75 and self.all_months:
            month = random.choice(self.all_months)
        else:
            month = f"{random.randint(1, 12):02d}"
        
        # === DIVERSE YEAR GENERATION ===
        r = random.random()
        if r < 0.25 and pattern.years:
            year = random.choice(pattern.years)
        elif r < 0.75 and self.all_years:
            year = random.choice(self.all_years)
        else:
            year = str(2026 + random.randint(0, 5))
        
        # Validate year is not expired
        try:
            year_int = int(year)
            if year_int < 2026:
                year = str(2026 + random.randint(0, 4))
        except:
            year = "2027"
        
        # Determine card type
        card_type = pattern.card_type or self._get_card_type(card_num)
        
        return {
            'card_number': card_num,
            'expiry_month': int(month),
            'expiry_year': int(year),
            'cvv': cvv,
            'card_type': card_type,
            'bin': pattern.prefix[:6]
        }
    
    def _get_card_type(self, card_number: str) -> str:
        """Detect card type from number"""
        if not card_number:
            return "UNKNOWN"
        
        first = card_number[0]
        if first == '4':
            return "Visa"
        elif first == '5':
            return "Mastercard"
        elif first == '3':
            return "American Express"
        elif first == '6':
            return "Discover"
        return "UNKNOWN"
    
    def get_sorted_patterns(self) -> List[PatternData]:
        """Get patterns sorted by quality score"""
        def score(p: PatternData) -> int:
            s = p.count * 10
            # Prefer VISA and Mastercard
            if p.card_type in ['Visa', 'Mastercard']:
                s += 20
            return s
        
        patterns = list(self.patterns.values())
        patterns.sort(key=score, reverse=True)
        return patterns
    
    def generate_cards(self, count: int) -> List[Dict]:
        """Generate specified number of cards"""
        if not self.patterns:
            print("[!] No patterns loaded. Call build_patterns first.")
            return []
        
        patterns = self.get_sorted_patterns()
        result = []
        
        # Distribute across patterns
        cards_per_pattern = max(1, count // len(patterns))
        
        for pattern in patterns:
            for _ in range(cards_per_pattern):
                card = self.generate_from_pattern(pattern)
                result.append(card)
                
                if len(result) >= count:
                    break
            
            if len(result) >= count:
                break
        
        # Fill remaining if needed
        while len(result) < count and patterns:
            pattern = random.choice(patterns)
            card = self.generate_from_pattern(pattern)
            result.append(card)
        
        # Shuffle
        random.shuffle(result)
        
        return result[:count]
    
    def generate_single(self) -> Optional[Dict]:
        """Generate a single card"""
        if not self.patterns:
            return None
        
        # Pick weighted random pattern (prefer higher count)
        patterns = list(self.patterns.values())
        weights = [p.count for p in patterns]
        total = sum(weights)
        weights = [w / total for w in weights]
        
        pattern = random.choices(patterns, weights=weights, k=1)[0]
        return self.generate_from_pattern(pattern)


# Singleton instance for use across the application
_generator_instance: Optional[SmartGenerator] = None


def get_generator(cards_file: str = "output/cards.json") -> SmartGenerator:
    """Get or create the singleton generator instance"""
    global _generator_instance
    
    if _generator_instance is None:
        _generator_instance = SmartGenerator(cards_file)
        cards = _generator_instance.load_cards()
        if cards:
            _generator_instance.build_patterns(cards)
    
    return _generator_instance


def refresh_patterns(cards_file: str = "output/cards.json"):
    """Reload patterns from updated card data"""
    global _generator_instance
    
    gen = SmartGenerator(cards_file)
    cards = gen.load_cards()
    if cards:
        gen.build_patterns(cards)
        _generator_instance = gen
        print(f"[OK] Refreshed patterns from {len(cards)} cards")


if __name__ == "__main__":
    # Test the generator
    print("\n" + "=" * 60)
    print("   Smart Pattern-Based Card Generator - Test")
    print("=" * 60)
    
    gen = SmartGenerator()
    cards = gen.load_cards()
    
    if cards:
        gen.build_patterns(cards)
        
        print("\n[+] Generating 10 sample cards...")
        generated = gen.generate_cards(10)
        
        for card in generated:
            cn = card['card_number']
            month = card['expiry_month']
            year = card['expiry_year']
            cvv = card['cvv']
            ct = card['card_type']
            print(f"   {cn}|{month:02d}|{year}|{cvv} [{ct}]")
    else:
        print("[!] No cards found to learn from.")

File: smart_generator.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

type Card struct {
	CardNumber string `json:"card_number"`
	Month      string `json:"month"`
	Year       string `json:"year"`
	CVV        string `json:"cvv"`
	CardType   string `json:"card_type"`
	Level      string `json:"level"`
	Bank       string `json:"bank"`
	Country    string `json:"country"`
}

type PatternData struct {
	Prefix          string
	SuffixDigits    [][]int       // All observed digits at each suffix position
	Months          []string      // All observed months
	Years           []string      // All observed years
	CVVs            []string      // All observed CVVs
	Bank            string
	Country         string
	CardType        string
	Level           string
	Count           int
	Cards           []Card
}

type GeneratedCard struct {
	CardNumber string `json:"card_number"`
	Month      string `json:"month"`
	Year       string `json:"year"`
	CVV        string `json:"cvv"`
	Expiry     string `json:"expiry"`
	CardType   string `json:"card_type"`
	Level      string `json:"level"`
	Bank       string `json:"bank"`
	Country    string `json:"country"`
	BIN        string `json:"bin"`
}

func main() {
	rand.Seed(time.Now().UnixNano())

	fmt.Println("╔════════════════════════════════════════════════════════════════╗")
	fmt.Println("║  🧠 SMART PATTERN-BASED CARD GENERATOR                         ║")
	fmt.Println("║  Reference real patterns + Diverse CVV/Expiry generation       ║")
	fmt.Println("╚════════════════════════════════════════════════════════════════╝")

	// Load real cards
	cards, err := loadCards("cards_realtime.json")
	if err != nil {
		fmt.Printf("❌ Error: %v\n", err)
		return
	}
	fmt.Printf("\n✅ Loaded %d working cards as reference\n", len(cards))

	// Build pattern database from real cards
	fmt.Println("\n🔬 Building pattern database from real cards...")
	patterns := buildPatternDatabase(cards)
	fmt.Printf("   Found %d unique patterns with 2+ cards\n", len(patterns))

	// Analyze real CVV and expiry distributions
	allCVVs, allMonths, allYears := analyzeDistributions(cards)
	fmt.Printf("   Unique CVVs: %d, Months: %d, Years: %d\n", len(allCVVs), len(allMonths), len(allYears))

	// Generate cards
	fmt.Println("\n💳 Generating smart pattern-based cards...")
	
	// Sort patterns by quality
	sortedPatterns := sortPatternsByQuality(patterns)

	// Generate different quantities
	gen500 := generateSmartCards(sortedPatterns, 500, allCVVs, allMonths, allYears)
	gen1000 := generateSmartCards(sortedPatterns, 1000, allCVVs, allMonths, allYears)
	gen2000 := generateSmartCards(sortedPatterns, 2000, allCVVs, allMonths, allYears)

	// Save outputs
	saveCards(gen500, "smart_500")
	saveCards(gen1000, "smart_1000")
	saveCards(gen2000, "smart_2000")

	// Print summary
	printSummary(gen1000)
}

func loadCards(filename string) ([]Card, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var cards []Card
	scanner := bufio.NewScanner(file)
	buf := make([]byte, 0, 64*1024)
	scanner.Buffer(buf, 1024*1024)

	for scanner.Scan() {
		var card Card
		if err := json.Unmarshal(scanner.Bytes(), &card); err != nil {
			continue
		}
		if len(card.CardNumber) >= 15 {
			cards = append(cards, card)
		}
	}
	return cards, scanner.Err()
}

func buildPatternDatabase(cards []Card) map[string]*PatternData {
	patterns := make(map[string]*PatternData)

	for _, card := range cards {
		cn := card.CardNumber
		if len(cn) < 16 {
			continue
		}

		// Use 8-digit prefix as pattern key
		prefix := cn[:8]
		suffix := cn[8:]

		if _, exists := patterns[prefix]; !exists {
			patterns[prefix] = &PatternData{
				Prefix:       prefix,
				SuffixDigits: make([][]int, 8),
			}
			for i := 0; i < 8; i++ {
				patterns[prefix].SuffixDigits[i] = []int{}
			}
		}

		p := patterns[prefix]
		p.Count++
		p.Cards = append(p.Cards, card)

		// Record suffix digits at each position
		for i, ch := range suffix {
			if i < 8 {
				digit := int(ch - '0')
				if digit >= 0 && digit <= 9 {
					p.SuffixDigits[i] = append(p.SuffixDigits[i], digit)
				}
			}
		}

		// Record month, year, cvv
		p.Months = append(p.Months, card.Month)
		p.Years = append(p.Years, card.Year)
		p.CVVs = append(p.CVVs, card.CVV)

		// Set metadata
		if p.Bank == "" && card.Bank != "" {
			p.Bank = card.Bank
			p.Country = card.Country
			p.CardType = card.CardType
			p.Level = card.Level
		}
	}

	// Filter to patterns with 2+ cards
	filtered := make(map[string]*PatternData)
	for k, v := range patterns {
		if v.Count >= 2 {
			filtered[k] = v
		}
	}

	return filtered
}

func analyzeDistributions(cards []Card) ([]string, []string, []string) {
	cvvSet := make(map[string]bool)
	monthSet := make(map[string]bool)
	yearSet := make(map[string]bool)

	for _, card := range cards {
		if len(card.CVV) >= 3 {
			cvvSet[card.CVV] = true
		}
		if len(card.Month) == 2 {
			monthSet[card.Month] = true
		}
		if len(card.Year) == 4 {
			yearSet[card.Year] = true
		}
	}

	cvvs := mapKeys(cvvSet)
	months := mapKeys(monthSet)
	years := mapKeys(yearSet)

	return cvvs, months, years
}

func mapKeys(m map[string]bool) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	return keys
}

func sortPatternsByQuality(patterns map[string]*PatternData) []*PatternData {
	var sorted []*PatternData
	for _, p := range patterns {
		if p.Bank != "" {
			sorted = append(sorted, p)
		}
	}

	sort.Slice(sorted, func(i, j int) bool {
		scoreI := calculatePatternScore(sorted[i])
		scoreJ := calculatePatternScore(sorted[j])
		return scoreI > scoreJ
	})

	return sorted
}

func calculatePatternScore(p *PatternData) int {
	score := p.Count * 10

	// Premium banks bonus
	bankUpper := strings.ToUpper(p.Bank)
	if strings.Contains(bankUpper, "CHASE") ||
		strings.Contains(bankUpper, "CITI") ||
		strings.Contains(bankUpper, "AMERICA") ||
		strings.Contains(bankUpper, "WELLS FARGO") ||
		strings.Contains(bankUpper, "CAPITAL ONE") {
		score += 100
	}

	// Premium countries
	if p.Country == "US" || p.Country == "CA" || p.Country == "GB" || p.Country == "UK" {
		score += 50
	}

	// Credit cards bonus
	if strings.Contains(strings.ToUpper(p.Level), "CREDIT") {
		score += 30
	}

	return score
}

func generateSmartCards(patterns []*PatternData, count int, allCVVs, allMonths, allYears []string) []GeneratedCard {
	var result []GeneratedCard

	// Distribute across patterns
	cardsPerPattern := 5
	patternsNeeded := count / cardsPerPattern
	if patternsNeeded > len(patterns) {
		patternsNeeded = len(patterns)
		cardsPerPattern = count / patternsNeeded
	}

	for i := 0; i < patternsNeeded; i++ {
		p := patterns[i%len(patterns)]

		for j := 0; j < cardsPerPattern; j++ {
			card := generateFromPattern(p, allCVVs, allMonths, allYears)
			result = append(result, card)

			if len(result) >= count {
				break
			}
		}

		if len(result) >= count {
			break
		}
	}

	// Shuffle
	rand.Shuffle(len(result), func(i, j int) {
		result[i], result[j] = result[j], result[i]
	})

	return result
}

func generateFromPattern(p *PatternData, allCVVs, allMonths, allYears []string) GeneratedCard {
	// Determine card length
	cardLen := 16
	cvvLen := 3
	if p.CardType == "AMEX" {
		cardLen = 15
		cvvLen = 4
	}
	suffixLen := cardLen - len(p.Prefix)

	// Generate suffix based on observed digit patterns
	suffix := ""
	for pos := 0; pos < suffixLen-1; pos++ { // -1 for Luhn digit
		if pos < len(p.SuffixDigits) && len(p.SuffixDigits[pos]) > 0 {
			// Pick random digit from observed digits at this position
			digits := p.SuffixDigits[pos]
			digit := digits[rand.Intn(len(digits))]
			suffix += strconv.Itoa(digit)
		} else {
			suffix += strconv.Itoa(rand.Intn(10))
		}
	}
	suffix += "0" // Placeholder for Luhn

	// Build card number
	cardNum := p.Prefix + suffix
	cardNum = makeLuhnValid(cardNum)

	// === DIVERSE CVV GENERATION ===
	// Option 1: Pick from this pattern's observed CVVs (30% chance)
	// Option 2: Pick from all observed CVVs (40% chance)
	// Option 3: Generate random CVV (30% chance)
	var cvv string
	r := rand.Float32()
	if r < 0.30 && len(p.CVVs) > 0 {
		// From pattern's CVVs
		cvv = p.CVVs[rand.Intn(len(p.CVVs))]
	} else if r < 0.70 && len(allCVVs) > 0 {
		// From all observed CVVs
		cvv = allCVVs[rand.Intn(len(allCVVs))]
	} else {
		// Random generate
		cvv = generateRandomDigits(cvvLen)
	}
	// Ensure correct length
	for len(cvv) < cvvLen {
		cvv = "0" + cvv
	}
	if len(cvv) > cvvLen {
		cvv = cvv[:cvvLen]
	}

	// === DIVERSE MONTH GENERATION ===
	// Option 1: Pick from this pattern's months (25% chance)
	// Option 2: Pick from all observed months (50% chance)
	// Option 3: Generate random month (25% chance)
	var month string
	r = rand.Float32()
	if r < 0.25 && len(p.Months) > 0 {
		month = p.Months[rand.Intn(len(p.Months))]
	} else if r < 0.75 && len(allMonths) > 0 {
		month = allMonths[rand.Intn(len(allMonths))]
	} else {
		month = fmt.Sprintf("%02d", rand.Intn(12)+1)
	}

	// === DIVERSE YEAR GENERATION ===
	// Option 1: Pick from this pattern's years (25% chance)
	// Option 2: Pick from all observed years (50% chance)
	// Option 3: Generate random valid year (25% chance)
	var year string
	r = rand.Float32()
	if r < 0.25 && len(p.Years) > 0 {
		year = p.Years[rand.Intn(len(p.Years))]
	} else if r < 0.75 && len(allYears) > 0 {
		year = allYears[rand.Intn(len(allYears))]
	} else {
		year = fmt.Sprintf("%d", 2026+rand.Intn(6))
	}

	// Validate year is not expired
	yearInt, _ := strconv.Atoi(year)
	if yearInt < 2026 {
		year = fmt.Sprintf("%d", 2026+rand.Intn(5))
	}

	return GeneratedCard{
		CardNumber: cardNum,
		Month:      month,
		Year:       year,
		CVV:        cvv,
		Expiry:     month + "/" + year,
		CardType:   p.CardType,
		Level:      p.Level,
		Bank:       p.Bank,
		Country:    p.Country,
		BIN:        p.Prefix[:6],
	}
}

func makeLuhnValid(cardNumber string) string {
	partial := cardNumber[:len(cardNumber)-1]
	checkDigit := calculateLuhnCheckDigit(partial)
	return partial + strconv.Itoa(checkDigit)
}

func calculateLuhnCheckDigit(partial string) int {
	sum := 0
	isSecond := true
	for i := len(partial) - 1; i >= 0; i-- {
		d := int(partial[i] - '0')
		if isSecond {
			d *= 2
			if d > 9 {
				d -= 9
			}
		}
		sum += d
		isSecond = !isSecond
	}
	return (10 - (sum % 10)) % 10
}

func generateRandomDigits(n int) string {
	digits := ""
	for i := 0; i < n; i++ {
		digits += strconv.Itoa(rand.Intn(10))
	}
	return digits
}

func saveCards(cards []GeneratedCard, prefix string) {
	// Simple format
	var simple []string
	for _, c := range cards {
		simple = append(simple, fmt.Sprintf("%s|%s|%s|%s", c.CardNumber, c.Month, c.Year, c.CVV))
	}
	filename := prefix + "_cards.txt"
	os.WriteFile(filename, []byte(strings.Join(simple, "\n")), 0644)
	fmt.Printf("   📁 %s (%d cards)\n", filename, len(cards))

	// Detailed format
	var detailed []string
	for _, c := range cards {
		detailed = append(detailed, fmt.Sprintf("%s|%s|%s|%s | %s | %s | %s | %s",
			c.CardNumber, c.Month, c.Year, c.CVV, c.CardType, c.Level, c.Bank, c.Country))
	}
	detailFile := prefix + "_detailed.txt"
	os.WriteFile(detailFile, []byte(strings.Join(detailed, "\n")), 0644)

	// JSON format
	jsonData, _ := json.MarshalIndent(cards, "", "  ")
	jsonFile := prefix + "_cards.json"
	os.WriteFile(jsonFile, jsonData, 0644)
}

func printSummary(cards []GeneratedCard) {
	fmt.Println("\n╔════════════════════════════════════════════════════════════════╗")
	fmt.Println("║  ✅ SMART GENERATION COMPLETE                                  ║")
	fmt.Println("╚════════════════════════════════════════════════════════════════╝")

	// Analyze diversity
	cvvSet := make(map[string]bool)
	monthSet := make(map[string]bool)
	yearSet := make(map[string]bool)
	bankSet := make(map[string]bool)
	typeCount := make(map[string]int)

	for _, c := range cards {
		cvvSet[c.CVV] = true
		monthSet[c.Month] = true
		yearSet[c.Year] = true
		if c.Bank != "" {
			bankSet[c.Bank] = true
		}
		typeCount[c.CardType]++
	}

	fmt.Printf("\n📊 Diversity Stats (from %d cards):\n", len(cards))
	fmt.Printf("   Unique CVVs:   %d (%.1f%% diversity)\n", len(cvvSet), float64(len(cvvSet))/float64(len(cards))*100)
	fmt.Printf("   Unique Months: %d\n", len(monthSet))
	fmt.Printf("   Unique Years:  %d\n", len(yearSet))
	fmt.Printf("   Unique Banks:  %d\n", len(bankSet))

	fmt.Println("\n🃏 By Card Type:")
	for t, c := range typeCount {
		fmt.Printf("   %-12s : %d\n", t, c)
	}

	fmt.Println("\n📁 Output Files:")
	fmt.Println("   smart_500_cards.txt      - 500 cards")
	fmt.Println("   smart_1000_cards.txt     - 1000 cards")
	fmt.Println("   smart_2000_cards.txt     - 2000 cards")
	fmt.Println("   + detailed.txt and .json versions")

	// Print samples showing diversity
	fmt.Println("\n📋 Sample Cards (showing CVV/Expiry diversity):")
	shown := make(map[string]bool)
	count := 0
	for _, c := range cards {
		key := c.BIN
		if !shown[key] && count < 15 {
			shown[key] = true
			count++
			fmt.Printf("   %s|%s|%s|%s [%s] %s\n",
				c.CardNumber, c.Month, c.Year, c.CVV, c.CardType, truncStr(c.Bank, 25))
		}
	}
}

func truncStr(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n-3] + "..."
}
